### PR TITLE
Making "upgrade project" work for async Main

### PR DESCRIPTION
**Pull request by Neme12**
Fixes https://github.com/dotnet/roslyn/issues/26326#issuecomment-383964525

Since the compiler only produces "Feature 'async main' is not available in C# 7.0" at emit time, there is no squiggle in the editor and it's not possible to use the upgrade project code fix to update the language version if anyone encounters the error.

This PR adds an IDE analyzer that tries to create its own version of this diagnostic so that the upgrade project feature works as expected. The diagnostic is produced if there's a Task-returning Main method with the right signature to be an entry point, we're in a compilation that requires an entry point, and there is no other entry point already. Its severity is a warning.

image

Tagging @jcouv

### CODE REVIEW

File: src/Features/CSharp/Portable/UpgradeProject/CSharpAsyncMainDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                IsValidAsyncMainReturnType(method.ReturnType, types) &&
+                method.DeclaringSyntaxReferences.Length == 1)
+            {
+                var declarationReference = method.DeclaringSyntaxReferences[0];
+
+                var parseOptions = (CSharpParseOptions)declarationReference.SyntaxTree.Options;
+                if (parseOptions.Kind == SourceCodeKind.Regular &&
+                    parseOptions.LanguageVersion < LanguageVersion.CSharp7_1)
+                {
+                    var entryPoint = context.Compilation.GetEntryPoint(context.CancellationToken);
```
Comment (Neme12) [inline src/Features/CSharp/Portable/UpgradeProject/CSharpAsyncMainDiagnosticAnalyzer.cs]: @CyrusNajmabadi  I think I remember you mentioning somewhere that this is really expensive. Is that the case? If so, should I do this in some other place such as compilation end action? Or is it so bad that we can't use it at all?

File: src/Features/CSharp/Portable/UpgradeProject/CSharpAsyncMainDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            new DiagnosticDescriptor(
+                IDEDiagnosticIds.AsyncMainNotAvailableDiagnosticId,
+                new LocalizableResourceString(
+                    nameof(CSharpFeaturesResources.Feature_async_main_is_not_available),
+                    CSharpFeaturesResources.ResourceManager, typeof(CSharpFeaturesResources)),
+                new LocalizableResourceString(
+                    nameof(CSharpFeaturesResources.Feature_async_main_is_not_available_in_CSharp_0_Please_use_language_version_7_1_or_greater),
+                    CSharpFeaturesResources.ResourceManager, typeof(CSharpFeaturesResources)),
+                DiagnosticCategory.Compiler,
+                DiagnosticSeverity.Warning,
```
Comment (Neme12) [inline src/Features/CSharp/Portable/UpgradeProject/CSharpAsyncMainDiagnosticAnalyzer.cs]: Should this be an error?

File: src/Features/Core/Portable/Diagnostics/Analyzers/IDEDiagnosticIds.cs [hunk 3]
```diff
@@ -86,6 +86,8 @@ internal static class IDEDiagnosticIds
 
         public const string UseCompoundAssignmentDiagnosticId = "IDE0054";
 
+        public const string AsyncMainNotAvailableDiagnosticId = "IDE0055";
```
Comment (Neme12) [inline src/Features/Core/Portable/Diagnostics/Analyzers/IDEDiagnosticIds.cs]: Should this be below "Analyzer error Ids" and have an ID of `IDE1009` to differentiate it from usual IDE features? I know that this is not actually an analyzer error, but the group below also contains other things that are not analyzer errors, such as "UnboundIdentifierId", which is really similar to this feature in that it tries to substitute the work of the compiler.

**Comment by CyrusNajmabadi** [comment]
> Since the compiler only produces "Feature 'async main' is not available in C# 7.0" at emit time,

Can we not just fix that?  Then we don't have to duplicate anything the compiler might do.  @jcouv ?

**Comment by jcouv** [comment]
> Can we not just fix the compiler?

I'm not exactly sure how to do this (we'd need to call `GetEntryPoint()` when we get the diagnostics), but it seems feasible. 
Looking at the logic behind `GetEntryPoint`, it relies on symbols which must have been bound for getting diagnostics anyways.
Also, since the `async main` feature was done, Cyrus optimized the name-lookup for members, so the search for methods called "Main" is now cheaper. It's still going to scan through all the types/members though (that's still a perf concern that we'd want to monitor).

I dug through async main PRs and issues, but couldn't find a record of a fundamental reason we couldn't do that, and I can't think of any.
Tagging @gafter in case he know.

Relates to https://github.com/dotnet/roslyn/issues/17259

**Comment by CyrusNajmabadi** [comment]
> Also, since the async main feature was done, Cyrus optimized the name-lookup for members, so the search for methods called "Main" is now cheaper. It's still going to scan through all the types/members though (that's still a perf concern that we'd want to monitor).

The symbol searching part of 'finding Main' should now be fairly fast.  We walk the decl table (whch is always kept up to date), and we just ask each TypeDecl's name-table (also kept up to date) "do you have a Main member".  So this is just a matter of walking each type-decl, and calling .Contains on its internal dictionary of Name->Member.

I'd be surprised if this was more than a few ms in practice even on compilations with many types.  (And it would be pretty easy to measure the cost of this on roslyn projects).

**Comment by Neme12** [comment]
Also related to #29698

**Comment by jcouv** [comment]
Yes, let's do that (investigate further on compiler side first). Thanks

**Comment by jasonmalinowski** [comment]
I've blocked the PR just so I don't get trigger happy with it. :smile:

**Comment by jasonmalinowski** [comment]
@jcouv And to confirm, this is on some discussion schedule somewhere we can track? I see bug #17259 isn't assigned to anybody.

**Comment by jcouv** [comment]
@jasonmalinowski This issue is not actively scheduled ("Unknown" milestone), but contributions are welcome. It would be awesome if @Neme12 took a stab at that alternative fix.

**Comment by Neme12** [comment]
@jcouv I just opened this because I got the impression that this can't be fixed in the compiler from your statements  in my issue:
> From my understanding, emitting the final binary (Emit API) is the only API that causes the entry point to be realized. So I don't see how this diagnostic could be produced in a lower layer...

and also because you liked my comment suggesting the approach I implemented here.

**Comment by CyrusNajmabadi** [comment]
> Sorry, I'm not a compiler dev 😄 If you think this should be fixed in the compiler, I can just close this then.

Anyone can contribute to the compiler :)   it just means a different group are necessary for sign-off :)

**Comment by Neme12** [comment]
Sure, anyone who is capable of doing that

**Comment by jcouv** [comment]
@Neme12 Although there remains a bit of uncertainty in my mind for solving this in the compiler layer, but it seems worth exploring more. 
Initially, I thought it would not be possible (too big of a perf hit) so the IDE approach seemed a good workaround/approximation. But Cyrus then presented the case for why we may be able to solve this in the compiler after all. Sorry for the mixed signals.

**Pull request closed without merging**